packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js

```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
 			return;
 		}
 
-		menuItemsRef.current = {};
+		const itemIndex = {};
+		menuItems.forEach( ( item ) => {
+			if ( ! itemIndex[ item.parent ] ) {
+				itemIndex[ item.parent ] = [];
+			}
+			itemIndex[ item.parent ].push( item );
+		} );
 
-		const innerBlocks = [];
+		menuItemsRef.current = {};
 
-		for ( const menuItem of menuItems ) {
-			const block = createBlockFromMenuItem( menuItem );
-			menuItemsRef.current[ block.clientId ] = menuItem;
-			innerBlocks.push( block );
-		}
+		const createMenuItemBlocks = ( items ) => {
+			const innerBlocks = [];
+			for ( const item of items ) {
+				let menuItemInnerBlocks = [];
+				if ( itemIndex[ item.id ] && itemIndex[ item.id ].length > 0 ) {
+					menuItemInnerBlocks = createMenuItemBlocks(
+						itemIndex[ item.id ]
+					);
+				}
+				const block = createBlockFromMenuItem(
+					item,
+					menuItemInnerBlocks
+				);
+				menuItemsRef.current[ block.clientId ] = item;
+				innerBlocks.push( block );
+			}
+			return innerBlocks;
+		};
 
+		const innerBlocks = createMenuItemBlocks( itemIndex[ 0 ] );
 		setBlocks( [ createBlock( 'core/navigation', {}, innerBlocks ) ] );
 	}, [ menuItems ] );
 
 	const saveBlocks = () => {
-		const { innerBlocks } = blocks[ 0 ];
-
-		for ( const block of innerBlocks ) {
-			const menuItem = menuItemsRef.current[ block.clientId ];
-
-			if ( ! menuItem ) {
-				saveMenuItem( {
-					...createMenuItemAttributesFromBlock( block ),
-					menus: menuId,
-				} );
-				continue;
+		const { clientId, innerBlocks } = blocks[ 0 ];
+
+		const saveNestedBlocks = ( nestedBlocks, parentId ) => {
+			for ( const block of nestedBlocks ) {
+				if ( block.innerBlocks && block.innerBlocks.length > 0 ) {
```

Comment:
I think that `createBlock()` always sets `innerBlocks`, so checking for that isn't necessary.

```js
if ( block.innerBlocks.length ) {
```